### PR TITLE
feat(core): shared libraries for problem handling

### PR DESCRIPTION
**Pull request by eirikurn**
## What

This commit implements the Problem Details specification for our projects.

### @island.is/nest/problem

* contains a NestJS module with filters that implement consistent problem responses.
* has error classes to manually throw problems.

### @island.is/shared/problem

* has TypeScript interfaces and types for all of our problems.
* utility to handle problems in GraphQL errors.

### Additional changes

* Integrated problems into application-system-api, api and application-system-form to better handle errors that occur in updateApplication and submitApplication.
* Refactored infra-nest-server to have opt-out support for ProblemModule but decided against it to keep the scope of this commit smaller. Instead it can be manually imported in your app's root module. Some of the refactor changes remain.

For more information, check out the RFC:
https://datatracker.ietf.org/doc/html/rfc7807

Fixes https://github.com/island-is/internal/issues/11

## Why

To follow our API design guide with consistent error handling and great DX.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/shared/problem/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+```typescript
+import { useMutation } from '@apollo/client'
+import { ProblemType, findProblemInApolloError } from '@island.is/shared/problem'
+
+// In some component:
+const [callApi, { error }] = useMutation(MUTATION)
+const problem = findProblemInApolloError(error) // undefined | Problem
+
+if (problem && problem.type === ProblemType.) {
```
Comment (sindrig) [inline libs/shared/problem/README.md:19]: `ProblemType.something`?

File: libs/shared/problem/src/ProblemType.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export enum ProblemType {
+  HTTP_BAD_REQUEST = 'https://httpstatuses.com/400',
+  HTTP_UNAUTHORIZED = 'https://httpstatuses.com/401',
+  HTTP_FORBIDDEN = 'https://httpstatuses.com/403',
+  HTTP_NOT_FOUND = 'https://httpstatuses.com/404',
+  HTTP_INTERNAL_SERVER_ERROR = 'https://httpstatuses.com/500',
+  VALIDATION_PROBLEM = 'https://docs.devland.is/reference/problems/validation-problem',
```
Comment (sindrig) [inline libs/shared/problem/src/ProblemType.ts:7]: I might have missed it, but where are you going to define this problem in the docs?

Comment (sindrig) [reply libs/shared/problem/src/ProblemType.ts:7]: It's also missing from the readme, that is, how to add the problem information to docs.devland.is for dereferencing

File: libs/nest/problem/src/lib/test/setup.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import { Controller, Get, Module } from '@nestjs/common'
+import request, { Test } from 'supertest'
+import { testServer } from '@island.is/infra-nest-server'
+import { GraphQLModule, Query, Resolver } from '@nestjs/graphql'
+import { Logger, LOGGER_PROVIDER } from '@island.is/logging'
+import { ProblemModule } from '../problem.module'
+
+export type CreateRequest = (mode?: 'rest' | 'graphql') => Test
```
Comment (baering) [inline libs/nest/problem/src/lib/test/setup.ts:8]: I noticed the `'graphql'` string is hardcoded in serveral places. Should it be an enum?

File: libs/nest/problem/src/lib/problem.options.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export const PROBLEM_MODULE_OPTIONS = 'PROBLEM_MODULE_OPTIONS'
+
+export interface ProblemModuleOptions {
+  convertAllErrors: boolean
```
Comment (baering) [inline libs/nest/problem/src/lib/problem.options.ts:4]: I don't think this is used anywhere

File: libs/nest/problem/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+throw new ProblemError({
+  type: ProblemType.HTTP_NOT_FOUND,
+  title: 'Not found',
+})
+```
+
+The following sub-classes are available for your convenience:
+
+- new ValidationError(fields) - accepts a record of validation issues.
```
Comment (saevarma) [inline libs/nest/problem/README.md:51]: `new ValidationProblem(fields)`

File: libs/nest/problem/src/lib/problem.filter.spec.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    )
+    expect(response.body).toMatchInlineSnapshot(`
+      Object {
+        "detail": "Found issues in these fields: field",
+        "fields": Object {
+          "field": "error",
+        },
+        "status": 400,
+        "title": "Validation failed",
+        "type": "https://docs.devland.is/reference/problems/validation-problem",
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/nest/problem/src/lib/problem.filter.spec.ts:32]: Not when using snapshots.

File: libs/nest/problem/src/lib/httpException.filter.spec.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    // Assert
+    expect(response.headers['content-type']).toContain(
+      'application/problem+json',
+    )
+    expect(response.body).toMatchInlineSnapshot(`
+      Object {
+        "status": 403,
+        "title": "Forbidden",
+        "type": "https://httpstatuses.com/403",
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/nest/problem/src/lib/httpException.filter.spec.ts:30]: Not when using snapshots.

File: libs/shared/problem/README.md (hunk 1 above)
Comment (eirikurn) [reply libs/shared/problem/README.md:19]: Good catch!

File: libs/nest/problem/src/lib/ProblemError.spec.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    const exception = new HttpException('Test', 500)
+
+    // Act
+    const problemError = ProblemError.fromHttpException(exception)
+
+    // Assert
+    expect(problemError).toMatchObject({
+      message: 'Test',
+      problem: {
+        type: 'https://httpstatuses.com/500',
```
[the remark this answers is not in the document]
Comment (saevarma) [reply libs/nest/problem/src/lib/ProblemError.spec.ts:54]: Forgot this?

File: libs/nest/problem/src/lib/apolloError.filter.spec.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    // Assert
+    expect(response.headers['content-type']).toContain(
+      'application/problem+json',
+    )
+    expect(response.body).toMatchInlineSnapshot(`
+      Object {
+        "detail": "User does not access.",
+        "status": 403,
+        "title": "Forbidden",
+        "type": "https://httpstatuses.com/403",
```
Comment (saevarma) [inline libs/nest/problem/src/lib/apolloError.filter.spec.ts:29]: Use appropriate ProblemType

File: libs/nest/problem/src/lib/problem.filter.spec.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    )
+    expect(response.body).toMatchInlineSnapshot(`
+      Object {
+        "detail": "Found issues in these fields: field",
+        "fields": Object {
+          "field": "error",
+        },
+        "status": 400,
+        "title": "Validation Failed",
+        "type": "https://docs.devland.is/reference/problems/validation-failed",
```
Comment (saevarma) [inline libs/nest/problem/src/lib/problem.filter.spec.ts:32]: Use appropriate ProblemType

File: libs/nest/problem/src/lib/ProblemError.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/nest/problem/src/lib/ProblemError.spec.ts:54]: Yup!

File: libs/nest/problem/src/lib/problem.filter.spec.ts (hunk 10 above)
Comment (eirikurn) [reply libs/nest/problem/src/lib/problem.filter.spec.ts:32]: This is a snapshot.

**Pull request merged**